store/sqlstore/post_store.go

```diff
@@ -1858,3 +1921,79 @@ func (s *SqlPostStore) GetOldestEntityCreationTime() (int64, error) {
 	}
 	return oldest, nil
 }
+
+func (s *SqlPostStore) cleanupThreads(postId, rootId, userId string) error {
+	if len(rootId) > 0 {
+		thread, err := s.Thread().Get(rootId)
+		if err != nil {
+			if err != sql.ErrNoRows {
+				return errors.Wrap(err, "failed to get a thread")
+			}
+		}
+		if thread != nil {
+			thread.ReplyCount -= 1
+			thread.Who = thread.Who.Remove(userId)
+			if _, err = s.Thread().Update(thread); err != nil {
+				return errors.Wrap(err, "failed to update thread")
+			}
+		}
+	}
+	_, err := s.GetMaster().Exec("DELETE FROM Threads WHERE PostId = :Id", map[string]interface{}{"Id": postId})
+	if err != nil {
+		return errors.Wrap(err, "failed to update Threads")
+	}
+	return nil
+}
+
+func (s *SqlPostStore) updateThreadsFromPosts(transaction *gorp.Transaction, posts []*model.Post) error {
+	postsByRoot := map[string][]*model.Post{}
+	for _, post := range posts {
+		// skip if post is not a part of a thread
+		if len(post.RootId) == 0 {
+			continue
+		}
+		postsByRoot[post.RootId] = append(postsByRoot[post.RootId], post)
+	}
+	now := model.GetMillis()
+	for rootId, posts := range postsByRoot {
+		var thread model.Thread
+		if err := transaction.SelectOne(&thread, "SELECT * from Threads WHERE PostId=:PostId", map[string]interface{}{"PostId": rootId}); err != nil {
```

Comment:
I think is not super important (probably only going to affect bulk import), but we could just select everything in one single query instead of one query per root post. But I think that is not critical, because is not in the critical path.